File: kovaak_tracker/coach/knowledge_active.py
"""Active knowledge resolution: the single entry point that decides whether
Coach reads the official first-party knowledge base or an installed pack.

Semantics (kb-sdk plan C3/C4): ``DATA_ROOT/config/knowledge.json`` selects the
active knowledge base. A missing config (or missing DATA_ROOT) is the default
official state; a broken config or an ``active`` pointer that cannot be
resolved degrades to official with a recorded fallback reason, and a pack
registry that fails to load degrades the same way. Official mapping assets
follow the same resource-root rule as ``knowledge_registry``; a missing
official mapping file is the normal built-in-fallback state, not an error.

This module never edits ``knowledge_registry``; it calls it through the module
attribute so test harnesses can pin ``knowledge_registry.load_registry``.
"""
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from . import knowledge_registry

CONFIG_SCHEMA_VERSION = "knowledge_config.v1"
ACTIVE_OFFICIAL = "official"

_RESOURCE_ROOT = os.environ.get("AIMING_COOKIE_RESOURCE_ROOT", "").strip()
_OFFICIAL_MAPPING_PATH = (
    Path(_RESOURCE_ROOT) / "knowledge" / "mapping" / "official.v1.json"
    if _RESOURCE_ROOT
    else Path(__file__).resolve().parents[2] / "knowledge" / "mapping" / "official.v1.json"
)


@dataclass
class ActiveKnowledge:
    mode: str                    # "official" | "pack"
    pack_id: str | None
    pack_version: str | None
    registry_path: Path | None   # pack mode: registry.json inside the pack
    mapping_path: Path | None    # pack mode: mapping.json inside the pack; official: resource-root path


_last_fallback_reason: str | None = None


def _record_fallback(reason: str) -> None:
    global _last_fallback_reason
    _last_fallback_reason = reason


def last_fallback_reason() -> str | None:
    """Fallback reason of the most recent resolve/load call (None = no fallback)."""
    return _last_fallback_reason


def _data_root() -> Path | None:
    raw = os.environ.get("DATA_ROOT", "").strip()
    return Path(raw) if raw else None


def _config_path() -> Path | None:
    root = _data_root()
    return root / "config" / "knowledge.json" if root else None


def _official() -> ActiveKnowledge:
    return ActiveKnowledge(
        mode="official",
        pack_id=None,
        pack_version=None,
        registry_path=None,
        mapping_path=_OFFICIAL_MAPPING_PATH,
    )


def _load_config(path: Path) -> dict[str, Any] | None:
    """None = no usable config (missing file is the default official state)."""
    try:
        raw_bytes = path.read_bytes()
    except OSError:
        return None
    try:
        doc = json.loads(raw_bytes)
    except (UnicodeDecodeError, json.JSONDecodeError):
        _record_fallback("knowledge config is not valid JSON")
        return None
    if not isinstance(doc, dict) or doc.get("schema_version") != CONFIG_SCHEMA_VERSION:
        _record_fallback("knowledge config schema_version is unsupported")
        return None
    return doc


def _installed_entry(config: dict[str, Any], pack_id: str) -> dict[str, Any] | None:
    installed = config.get("installed")
    if not isinstance(installed, list):
        return None
    for item in installed:
        if isinstance(item, dict) and item.get("pack_id") == pack_id:
            return item
    return None


def _resolve() -> ActiveKnowledge:
    config_path = _config_path()
    if config_path is None:
        return _official()
    config = _load_config(config_path)
    if config is None:
        return _official()
    active = config.get("active")
    if not isinstance(active, str) or not active:
        _record_fallback("knowledge config active field is invalid")
        return _official()
    if active == ACTIVE_OFFICIAL:
        return _official()
    entry = _installed_entry(config, active)
    pack_dir = _data_root() / "knowledge-packs" / active
    if entry is None or not pack_dir.is_dir():
        _record_fallback(f"active knowledge pack is not installed: {active}")
        return _official()
    registry_path = pack_dir / "knowledge" / "registry.json"
    if not registry_path.is_file():
        _record_fallback(f"active knowledge pack registry is missing: {active}")
        return _official()
    mapping_path = pack_dir / "mapping.json"
    pack_version = entry.get("pack_version")
    return ActiveKnowledge(
        mode="pack",
        pack_id=active,
        pack_version=pack_version if isinstance(pack_version, str) else None,
        registry_path=registry_path,
        mapping_path=mapping_path if mapping_path.is_file() else None,
    )


def resolve_active() -> ActiveKnowledge:
    """Resolve the current active knowledge base; degraded states fall back to official."""
    global _last_fallback_reason
    _last_fallback_reason = None
    return _resolve()


def load_active_registry() -> dict[str, Any]:
    """Load the active registry; any pack failure falls back to official + reason."""
    global _last_fallback_reason
    _last_fallback_reason = None
    active = _resolve()
    if active.registry_path is None:
        return knowledge_registry.load_registry()
    try:
        return knowledge_registry.load_registry(path=active.registry_path)
    except Exception:
        _record_fallback(f"active knowledge pack registry is invalid: {active.pack_id}")
        return knowledge_registry.load_registry()


def load_active_mapping() -> tuple[dict[str, Any] | None, str | None]:
    """(mapping doc, fallback reason). A missing file -> (None, None): the
    built-in frozen rules are the normal fallback, not a degradation."""
    global _last_fallback_reason
    _last_fallback_reason = None
    active = _resolve()
    path = active.mapping_path
    if path is None:
        return (None, None)
    try:
        raw_bytes = path.read_bytes()
    except OSError:
        return (None, None)
    try:
        doc = json.loads(raw_bytes)
    except (UnicodeDecodeError, json.JSONDecodeError):
        reason = "active mapping is not valid JSON"
        _record_fallback(reason)
        return (None, reason)
    if not isinstance(doc, dict):
        reason = "active mapping is not a JSON object"
        _record_fallback(reason)
        return (None, reason)
    return (doc, None)

File: webapp/coach-runtime/src/knowledge-active.ts
/**
 * Active knowledge resolution for the TS/sidecar side (kb-sdk plan C4).
 *
 * DATA_ROOT/config/knowledge.json selects the active knowledge base, mirroring
 * the Python-side knowledge_active module. A missing config is the default
 * official state (not a fallback); a broken config or an `active` pointer that
 * cannot be resolved degrades to official with a recorded fallback reason, and
 * a pack registry that fails to load degrades the same way with a
 * console.error so the sidecar startup log shows why the official base came
 * back. TS never reads or evaluates mapping documents — registry only.
 */
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

import { getDataRoot } from "./app-data.ts";
import { isRecord } from "./contracts.ts";
import {
  loadKnowledgeRegistry,
  loadKnowledgeRegistryFromPath,
  type KnowledgeRegistry,
} from "./knowledge-registry.ts";

const CONFIG_SCHEMA_VERSION = "knowledge_config.v1";
const ACTIVE_OFFICIAL = "official";

export type ActiveKnowledge =
  | { mode: "official" }
  | { mode: "pack"; packId: string; registryPath: string };

let lastFallbackReason: string | null = null;

function recordFallback(reason: string): void {
  lastFallbackReason = reason;
}

/** Fallback reason of the most recent resolve/load call (null = no fallback). */
export function lastActiveKnowledgeFallbackReason(): string | null {
  return lastFallbackReason;
}

function readConfig(): Record<string, unknown> | null {
  const path = join(getDataRoot(), "config", "knowledge.json");
  let raw: string;
  try {
    raw = readFileSync(path, "utf-8");
  } catch {
    return null; // Missing config is the default official state, not a fallback.
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    recordFallback("knowledge config is not valid JSON");
    return null;
  }
  if (!isRecord(parsed)) {
    recordFallback("knowledge config is not a JSON object");
    return null;
  }
  if (parsed.schema_version !== CONFIG_SCHEMA_VERSION) {
    recordFallback("knowledge config schema_version is unsupported");
    return null;
  }
  return parsed;
}

function installedPackIds(config: Record<string, unknown>): Set<string> {
  const installed = config.installed;
  if (!Array.isArray(installed)) return new Set();
  return new Set(
    installed
      .filter(isRecord)
      .map((item) => item.pack_id)
      .filter((packId): packId is string => typeof packId === "string"),
  );
}

export function resolveActiveKnowledge(): ActiveKnowledge {
  lastFallbackReason = null;
  const config = readConfig();
  if (!config) return { mode: "official" };
  const active = config.active;
  if (typeof active !== "string" || active.length === 0) {
    recordFallback("knowledge config active field is invalid");
    return { mode: "official" };
  }
  if (active === ACTIVE_OFFICIAL) return { mode: "official" };
  const packDir = join(getDataRoot(), "knowledge-packs", active);
  if (!installedPackIds(config).has(active) || !existsSync(packDir)) {
    recordFallback(`active knowledge pack is not installed: ${active}`);
    return { mode: "official" };
  }
  const registryPath = join(packDir, "knowledge", "registry.json");
  if (!existsSync(registryPath)) {
    recordFallback(`active knowledge pack registry is missing: ${active}`);
    return { mode: "official" };
  }
  return { mode: "pack", packId: active, registryPath };
}

export function loadActiveKnowledgeRegistry(): KnowledgeRegistry {
  lastFallbackReason = null;
  const active = resolveActiveKnowledge();
  if (active.mode === "official") return loadKnowledgeRegistry();
  try {
    return loadKnowledgeRegistryFromPath(active.registryPath);
  } catch (error) {
    recordFallback(`active knowledge pack registry is invalid: ${active.packId}`);
    console.error(
      `[coach] active knowledge pack registry is invalid (${active.packId}), `
      + `falling back to official: ${error instanceof Error ? error.message : String(error)}`,
    );
    return loadKnowledgeRegistry();
  }
}

/** display_name of an installed pack from the active config (Coach source attribution). */
export function activePackDisplayName(packId: string): string | undefined {
  const config = readConfig();
  if (!config || !Array.isArray(config.installed)) return undefined;
  const entry = config.installed.find((item) => isRecord(item) && item.pack_id === packId);
  if (!entry || !isRecord(entry)) return undefined;
  const displayName = entry.display_name;
  return typeof displayName === "string" && displayName.length > 0 ? displayName : undefined;
}
